### PR TITLE
Fixes #22915 - Make the cleaner clean orphaned dynflow plans

### PR DESCRIPTION
**Pull request by adamruzicka**
TODO:
- [ ] i18n support

### CODE REVIEW

File: test/unit/cleaner_test.rb [hunk 1]
```diff
@@ -18,6 +19,8 @@ class TasksTest < ActiveSupport::TestCase
                          end,
                          FactoryBot.create(:dynflow_task, :product_create_task)]
       cleaner.expects(:tasks_to_csv)
+      # Hack to make the tests pass due to ActiveRecord shenanigans
+      cleaner.expects(:delete_orphaned_dynflow_tasks)
```
Comment (adamruzicka) [inline test/unit/cleaner_test.rb]: `ActiveRecord` doesn't actually save things to the db in tests, so the tests fail when accessing AR data using sequel.

File: lib/foreman_tasks/cleaner.rb [hunk 2]
```diff
@@ -149,21 +182,26 @@ def prepare_filter
...
-        @total = tasks.size
+        @total = total
         say "#{@current}/#{@total}", false
       end
     end
 
     def report_progress(chunk)
       if verbose
-        @current += chunk.size
+        @current += chunk.count
```
Comment (iNecas) [inline lib/foreman_tasks/cleaner.rb]: When testing, I was seeing

```
About to remove 3 orphaned execution plan(s)
Deleted  orphaned execution plan(s)
```
It seems the `@current` was not set in non-berbose mode. I guess, we should track the `@current` always, and only prevent the actual printing in the `verbose` mode.

File: lib/foreman_tasks/cleaner.rb [hunk 3]
```diff
@@ -138,8 +135,44 @@ def with_backup_file(backup_dir, file_name)
...
+      if noop
+        say '[noop] ' + noop_message if noop_message
+        say "[noop] #{source.count} #{name}(s) would be deleted"
+      else
+        yield source, name
+      end
+    end
+
+    def with_batches(source, name)
+      start_tracking_progress(name, source.count)
```
Comment (iNecas) [inline lib/foreman_tasks/cleaner.rb]: Currently, when having an empty set, I get:

```
About to remove 0 orphaned execution plan(s)
Deleted  orphaned execution plan(s)
```

I guess, in this case, we should just mention that no orphaned execution plans were found on go on.

**Comment by iNecas** [comment]
Thanks for coming up with this one. Few comments from testing, but looks promising in general.

There are some cleaner tests failures, that will probably be related.

File: lib/foreman_tasks/cleaner.rb (hunk 3 above)
Comment (adamruzicka) [reply lib/foreman_tasks/cleaner.rb]: Hope this is better
```
2018-03-16T14:54:25 [I|for|] No tasks found, skipping.                                                                                                                                                                                      
2018-03-16T14:54:25 [I|for|] No orphaned execution plans found, skipping.                                                                                                                                                                   
```

**Comment by adamruzicka** [comment]
[test]

**Comment by iNecas** [comment]
Running tests again http://ci.theforeman.org/job/test_plugin_pull_request/4796/

File: lib/foreman_tasks/cleaner.rb [hunk 4]
```diff
@@ -138,8 +135,49 @@ def with_backup_file(backup_dir, file_name)
...
+        say "[noop] #{source.count} #{name}(s) would be deleted"
+      else
+        yield source, name
+      end
+    end
+
+    def with_batches(source, name)
+      count = source.count
+      if count.zero?
+        say("No #{name}s found, skipping.")
```
Comment (iNecas) [inline lib/foreman_tasks/cleaner.rb]: Passing the arguments as plural + not having to use the additional `s` would be probably a bit more flexible. When I was testing, I got message:

```
No tasks found, skipping
```

It was not clear what tasks those were. Having `name` something like `tasks matching filter` would be more self-describing.

File: lib/foreman_tasks/cleaner.rb [hunk 5]
```diff
@@ -149,19 +187,20 @@ def prepare_filter
...
-      if verbose
-        @current += chunk.size
-        say "#{@current}/#{@total}", false
-      end
+      @current += chunk.count
+      say "#{@current}/#{@total}", false if verbose
+    end
+
+    def report_done(name)
+      say "Deleted #{@current} #{name}(s)"
```
Comment (iNecas) [inline lib/foreman_tasks/cleaner.rb]: With orphaned tasks, I was getting `Deleted 0 orphaned execution plans` even though there has been reported (and deleted` 1 orhphaned execution plan to be deleted

**Comment by adamruzicka** [comment]
Both comments should be addressed

**Comment by iNecas** [comment]
Tests on sqlite still seem to not be happy http://ci.theforeman.org/job/test_plugin_pull_request/4796/, ACK otherwise

File: test/unit/cleaner_test.rb [hunk 6]
```diff
@@ -5,6 +5,8 @@ class TasksTest < ActiveSupport::TestCase
     # To stop dynflow from backing up actions, execution_plans and steps
     ForemanTasks.dynflow.world.persistence.adapter.stubs(:backup_to_csv)
     ForemanTasks::Cleaner.any_instance.stubs(:say) # Make the tests silent
+    # Hack to make the tests pass due to ActiveRecord shenanigans
```
Comment (iNecas) [inline test/unit/cleaner_test.rb]: I didn't know `shenanigans` is actually a word. Definitely a nice one :)

**Comment by iNecas** [comment]
Thanks @adamruzicka

**Pull request merged**